Add more granular breakdown of work items required to build initial version of HLSL clang compiler c library (#200)
microsoft/hlsl-specs: proposals/0016-assets/compile-compute-shader-breakdown.md

<!-- {% raw %} -->

# C interface for HLSL compiler as a library - Work item breakdown #1
## (Adoption by ML team and ability to compile compute shader using library)

* Parent Proposal: [0016](../0016-c-interface-compiler-library.md)
* Author(s): [Cooper Partin](https://github.com/coopp)
* Sponsor: [Cooper Partin](https://github.com/coopp)
* Status: **Under Consideration**
* Impacted Project(s): (Clang)

## Introduction

An effort is underway to modernize compilation of HLSL based shaders by adding
HLSL compilation to clang. This document is a sub proposal for 
[0016](../0016-c-interface-compiler-library.md) to break down the work
required to achieve the first milestone of adopotion by the ML team and
ability to compile a compute shader using clang as a library.

> The Issues created here will be ordered to help contruct a roadmap that can be

Reviewer: ```suggestion
The Issues created here will be ordered to help construct a roadmap that can be
```

leveraged towards defining a realistic timeline to completion.

## Issues

> Implementation work in this list is assumed to include tests!

Reviewer: TBH, I think that some of the steps here could likely be combined, but this glosses over the bigger roadmap questions of what the entire DXC API surface covers.

> [] Design and spec where the library will be built/exposed in the llvm-project

Reviewer 2: Sadly these check boxes don't render in GitHub's markdown previews 😞

   source tree and how a developer would consume the library.

[] Design and spec the minimal entrypoints/function signatures and types
   for compiling a shader.

[] Validate design with ML team adopters to ensure that supported functionality
   being built meets their needs.
   The following discussion will help drive if the additional feartures will need
   to be added:
   * Add include handler support?
   * Expose linker support? (Might be included with clang from other work)

>    * Expose optmizer support? (Might be include with clanga from other work)

Reviewer: ```suggestion
   * Expose optimizer support? (Might be included with clang from other work)
```

>    * Expose tooling featrures like container reflection, and other output inspection

Reviewer: ```suggestion
   * Expose tooling features like container reflection, and other output inspection
```

>    * Expose tooling featrures like container reflection, and other output inspection

Reviewer: Somewhere / at some point, I think we need to generate the full list of features that the current API supports.

Author: This PR https://github.com/microsoft/hlsl-specs/pull/197 is an attempt to capture that

Reviewer: oops, I missed that!

     features?

> [] Add skeleton library to llvm-project build system. This includes adding code

Reviewer: What does "skeleton library" mean here?  Does this mean that there are non-functional entry points?  Or does it mean that not all the expected entry points are there?

   that builds/links/calls into to the library ensuring that it is configured
   properly.

> [] Add documentation for defined library entrypoints to llvm-project docs with

Reviewer: I'd argue that documentation should be assumed to be part of the implementation, along with tests.

   information on how to use/compile a HLSL shader and pass outputs to the 
   D3D12 runtime.
   * Is there a way to put these docs under an 'experimental' grouping?

> [] Design and spec out-of-process compile library architecture.

Reviewer: IMO this should be one of the first things you do - or at least got agreement on whether or not there's an in-proc version of this.  If it is out-of-proc only then I think building an in-proc interim version would be just building tech debt.
